### ISSUE TITLE
Directories on an already-mounted sshfs path show empty if they contain a symlink pointing at a non-existent file

### ISSUE THREAD

**Issue by satu** [microsoft/vscode#115645]
Issue Type: <b>Bug</b>

I can reliably reproduce on my Mac (Big Sur), in which I have a sshfs mount to  a Linux workstation.

1. Create a folder in the remote hdd, add some files and folders
2. Open the folder in VSCode and verify it works OK
3. Close folder in VSCode
4. Add a symlink in the remote hdd pointing to a non existing file, e.g. `ln -s myfile ../nonexistent`
5. Open the folder in VSCode

Expected result:
- folder can be open and navigated nicely

Observed result:
- folder looks empty in VSCode

Also tested in VSCode insider with same effect.

VS Code version: Code 1.52.1 (ea3859d4ba2f3e577a159bc91e3074c5d85c0523, 2020-12-16T16:30:02.420Z)
OS version: Darwin x64 20.2.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz (12 x 2600)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|3, 3, 3|
|Memory (System)|16.00GB (0.36GB free)|
|Process Argv|--crash-reporter-id e7ddb06c-71d4-4f10-be25-e8037994cb8b|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (4)</summary>

Extension|Author (truncated)|Version
---|---|---
android-system-tools|fli|0.0.2
vscode-sshfs|Kel|1.19.1
remote-vscode|raf|1.1.0
rewrap|stk|1.13.0

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
openlogontheside:30221877
python383:30185418
pythonvspyt617:30244266
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492:30211401
wsl2promptcf:30224613
pythonvsdeb440:30248342
unusedprompt:30224610
folderexplorercf:30224615
openfilemenucf:30224648
pythonvsded773:30248341
pythonvspyt875:30251590

```

</details>

**Comment by satu** [comment]
This seems very similar to #78743 which was closed as fixed but it's what lead me to find the bug.

**Comment by isidorn** [comment]
We are aware of some of the symlink limitation
fyi @bpasero for potential duplicate issue

**Comment by bpasero** [comment]
I would not expect the entire explorer to be empty, I think we should have code in place to protect from this, but maybe we miss a case.

Please see if you can find any logged output that could help us find the source of the issue:
* run `code --verbose` from the command line and attach the output
* in Code select **Help | Toggle Developer Tools** and select the **Console**
* attach any output you see in the **Console** you can find

Thanks!

**Comment by satu** [comment]
I find errors related to the incrimin
Found some errors related to the symlink:

From the console:

```
log.ts:173 TRACE EntryNotFound (FileSystemError): Error: ENOENT: no such file or directory, lstat '/usr/local/mysshfs/home/dariofreni/test/test123/non'
```

On the stdout:

```

log.ts:173 TRACE EntryNotFound (FileSystemError): Error: ENOENT: no such file or directory, lstat '/usr/local/mysshfs/home/dariofreni/test/test123/non'
dariofreni-macbookpro2:Desktop dariofreni$ grep test123/non ~/Desktop/vscode_verbose.log
[38434:0203/143701.019225:INFO:CONSOLE(1779)] "%cTRACE color: #888 EntryNotFound (FileSystemError): Error: ENOENT: no such file or directory, lstat '/usr/local/mysshfs/home/dariofreni/test/test123/non'", source: file:///Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js (1779)
```

full files attached.






[consoleout.txt](https://github.com/microsoft/vscode/files/5919185/consoleout.txt)
[vscode_verbose.log](https://github.com/microsoft/vscode/files/5919197/vscode_verbose.log)

**Comment by satu** [comment]
the directory setup I'm testing it is the following

```
directory test    <--- this is the one I'm opening in VSCode
  directory test123
    file test1.txt
    symlink non -> ../nonexistent     <--- this is the broken symlink
  file test.txt
```

I can see `directory test` and `file test.txt`, and I see `directory test123` but it's empty.

While harvesting information, I noticed that when I use `find` / `ls` on my Mac terminal on that directory, the non existent symlink is not even presented. So something is not right. It seems VSCode spots the file via syscall but then it's non-existent in other syscalls? I think we're onto something.

As far as I know sshfs tries to be clever and materialises symlinks as normal files. This is afaik default behaviour and I actually take advantage of this in my workflow. I sshfs-mount my home in my Linux remote machine, which contains a symlink called `src` pointing at an additional hdd mounted in `/mnt/extradrive`: on my Mac, I don't see this as symlink but as a normal folder that I can navigate and use.

**Comment by satu** [comment]
(to add: I also tried to open VSCode on the Linux machine via remote desktop and everything is shown, including the dead symlink)

**Comment by bpasero** [comment]
@satu oh, so the explorer is in fact only empty for that particular folder but not entirely? Then I misunderstood this issue...

I would have expected that you see all files of that folder that are valid and the dangling symbolic link would simply not show. Any chance you could try our insiders build to see if that makes any difference?

You can give our preview releases a try from: https://code.visualstudio.com/insiders/

@isidorn not sure what the explorer is doing, is it maybe calling `resolve` for each child? I guess not right, because we have this code that explicitly protects from errors like this one:

https://github.com/microsoft/vscode/blob/31c034604e5e70366b335c743ccdfb5eea6b7408/src/vs/platform/files/common/fileService.ts#L239-L250

**Comment by satu** [comment]
correct: the subdirectory test123 is shown as empty.

Note that when opening the test directory locally on Linux, test123 subdirectory shows both the dangling symlink and the regular file in there. If I click on the dangling symlink a dialog informs me "Unable to open 'non': Unable to read file.

This seems to be different from your expectation.

Going to download and try the insiders version.

**Comment by satu** [comment]
Same behaviour on insiders. See attached. Relevant portion:

```
$ grep test123/non vscode_insiders_*
vscode_insiders_console.log:workbench.desktop.main.js:sourcemap:618 TRACE EntryNotFound (FileSystemError): Error: ENOENT: no such file or directory, lstat '/usr/local/mysshfs/home/dariofreni/test/test123/non'
vscode_insiders_stdout.log:[44501:0203/160508.895395:INFO:CONSOLE(618)] "%cTRACE color: #888 EntryNotFound (FileSystemError): Error: ENOENT: no such file or directory, lstat '/usr/local/mysshfs/home/dariofreni/test/test123/non'", source: file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js (618)
```
[vscode_insiders_console.log](https://github.com/microsoft/vscode/files/5919819/vscode_insiders_console.log)
[vscode_insiders_stdout.log](https://github.com/microsoft/vscode/files/5919820/vscode_insiders_stdout.log)

**Comment by isidorn** [comment]
@bpasero we will resolve all the way down to the open file. However since @satu sees it with a clean insiders, I do no think there is a file open to resolve down to. So we should only resolve the top folder on start.

**Comment by bpasero** [comment]
I still do not understand it and believe the node.js library returns invalid stuff back to us. Distilling the stacks:

```
workbench.desktop.main.js:sourcemap:618 TRACE EntryNotFound (FileSystemError): Error: ENOENT: no such file or directory, lstat '/usr/local/mysshfs/home/dariofreni/test/test123/non'
    at Object.h [as createFileSystemProviderError] (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:614:10355)
    at D.toFileSystemProviderError (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:618:47233)
    at D.readdir (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:618:40854)
    at async o.toFileStat (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:618:25159)
    at async o.resolve (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:618:24106)
    at async g.fetchChildren (file:///Applications/Visual Studio Code - Insiders.app/Cont
```

This is:
* the explorer fetching children of a folder
* via `resolve` in the file service
* we think this is a folder (even though it seems to be a dangling symbolic link)
* we call `readdir`
* an exception is thrown

=> here I would have expected our detection of dangling symbolic links to prevent this error

```
/Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:618 TRACE EntryNotFound (FileSystemError): Error: ENOENT: no such file or directory, stat '/usr/local/mysshfs/home/dariofreni/test/node_modules'
    at Object.h [as createFileSystemProviderError] (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:614:10355)
    at D.toFileSystemProviderError (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:618:47233)
    at D.stat (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:618:40528)
    at async o.doResolveFile (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:618:24544)
    at async file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:618:25537
    at async Promise.all (index 0)
```

This is:
* probably `resolveAll`
* via `doResolveFile` (which is `Promise.all`)
* via `stat`
* resulting in a ENOENT

=> again this is weird if we can detect a dangling symbolic link

@satu do you have node.js installed on the command line and conduct a little experiment? I would be curious what the result of `fs.statSync` and `fs.lstatSync` of `/usr/local/mysshfs/home/dariofreni/test/test123/non` and `/usr/local/mysshfs/home/dariofreni/test/node_modules` returns. How do these folders map back to the hierarchy you presented here?

**Comment by satu** [comment]
I'm not familiar with node.js, sorry. And I don't have a `node_modules` dir in my setup.

I found a way to reproduce this locally on a Linux machine (should probably work on Mac too). I can confirm that the behaviour only happens when specifying the option `-o follow_symlinks` of the `sshfs` driver -- I found out that this was used in my work configuration and as I mentioned it's actually convenient to me.

Here is a script to repro locally provided that you have a `sshd` running:

```sh
cd ${HOME}
mkdir -p sshfs_exported/topdir/subdir
echo foo > sshfs_exported/topdir/foo.txt
echo bar > sshfs_exported/topdir/subdir/bar.txt
ln -s ../nonexistent sshfs_exported/topdir/subdir/broken_symlink

code sshfs_exported/topdir  # this opens vscode on local dirs, everything works.

# close vscode

# sudo apt install sshfs

mkdir sshfs_mountpoint
sshfs -o follow_symlinks -d ${USER}@localhost:${HOME}/sshfs_exported sshfs_mountpoint

code sshfs_mountpoint/topdir  # this opens vscode on sshfs-mounted dir, where the problem happens
```

**Comment by bpasero** [comment]
@satu excellent steps, I was able to reproduce locally and see the issue: when encountering a folder we call [`readdir` ](https://man7.org/linux/man-pages/man3/readdir.3.html) and this returns the following entries: `bar.txt` and `broken_symlink`. We then go ahead and [`lstat`](https://man7.org/linux/man-pages/man2/lstat.2.html) each child. When we do so for `broken_symlink` we unexpectedly get a `ENOENT` error which I would never expect when using `lstat` because that will ignore the fact that a symlink might point to an invalid target.

In other words: if `readdir` returns entries, `lstat` must always be able to run on them.

As such I think something is broken in the file system implementation of the SSH filesystem or even how node handles it.

I have pushed a workaround on our end to have a fallback and skip over the faulty directory entry.

![image](https://user-images.githubusercontent.com/900690/106899171-54c2b800-66f5-11eb-8359-842cde2aa6a5.png)

**Comment by bpasero** [comment]
Actually pushed a better solution because we do have support to show entries that are unknown and that way you get at least a correct view into the folder:

![image](https://user-images.githubusercontent.com/900690/106918633-a83f0100-6709-11eb-9a34-7982eb0b2f67.png)
